File: frontend/src/api/transaction.ts
import { get, post } from './client';
import type { Transaction, TransactionLeg, TransactionType } from '@/types';

export interface CreateTransactionRequest {
  occurredAt: string;
  type: TransactionType;
  settleDate?: string;
  note?: string;
  tags?: string[];
  legs: Omit<TransactionLeg, 'id'>[];
}

export interface TransactionListParams {
  from?: string;
  to?: string;
  type?: TransactionType;
  limit?: number;
  cursor?: string;
}

export const transactionApi = {
  // List transactions for portfolio
  list: (portfolioId: string, params?: TransactionListParams) =>
    get<Transaction[]>(`/v1/portfolios/${portfolioId}/transactions`, params),

  // Get transaction by ID
  getById: (id: string) =>
    get<Transaction>(`/v1/transactions/${id}`),

  // Create transaction
  create: (portfolioId: string, data: CreateTransactionRequest) =>
    post<Transaction>(`/v1/portfolios/${portfolioId}/transactions`, data),

  // Void transaction
  void: (id: string) =>
    post<Transaction>(`/v1/transactions/${id}/void`),
};

File: backend/src/main/java/com/portfolio/ledger/entity/Transaction.java
package com.portfolio.ledger.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "transactions")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Column(name = "portfolio_id", nullable = false)
    private String portfolioId;

    @Column(name = "occurred_at", nullable = false)
    private LocalDateTime occurredAt;

    @Column(name = "settle_date")
    private LocalDate settleDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionType type;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private TransactionStatus status = TransactionStatus.POSTED;

    @Column(columnDefinition = "TEXT")
    private String note;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    @Builder.Default
    private List<String> tags = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "transaction", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<TransactionLeg> legs = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public void addLeg(TransactionLeg leg) {
        legs.add(leg);
        leg.setTransaction(this);
    }

    public enum TransactionType {
        BUY,
        SELL,
        DIVIDEND,
        INTEREST,
        FEE,
        TAX,
        DEPOSIT,
        WITHDRAW,
        FX_CONVERT,
        SPLIT,
        MERGER,
        TRANSFER
    }

    public enum TransactionStatus {
        POSTED,
        VOID,
        PENDING
    }
}
